### PR TITLE
feat: add table.createInsertStream for native streaming inserts

### LINKED ISSUE #506: Implement writable streams for natively streaming inserts
**Is your feature request related to a problem? Please describe.**
It would be really useful to be able to natively stream records directly into a table (via `/insertAll`). This would make it much more accessible to transfer significant amounts of data on low-resource platforms like cloud functions.

 **Describe the solution you'd like**
A Table method similar to [createWriteStream()](https://googleapis.dev/nodejs/storage/latest/File.html#createWriteStream) in the @google-cloud/storage package. Being able to just call `mySourceStream.pipe(myTable.createWriteStream())` would be super helpful.

 **Describe alternatives you've considered**
I've managed to get roughly the functionality I'm looking for by calling `insert()` in arbitrarily sized batches, i.e.

```js
class TableWriteableStream extends Writable {
  constructor (destinationTable) {
    super({ objectMode: true })
    this.destinationTable = destinationTable
    this.records = []
  }

  _write (chunk, encoding, callback) {
    this.records.push(chunk)
    // Every 500 records, call insert()
    if (this.records.length >= 500) {
      const recordsToInsert = this.records
      this.records = []
      this.destinationTable
        .insert(recordsToInsert)
        .then(() => { callback() })
    } else {
      callback()
    }
  }

  _final (callback) {
      // Insert any remaining records ...

```

 **Additional context**
This is somewhat related to #75, though that seems focused on compression (which would be a great option for this feature, just like the 'gzip' option for the @google-cloud/storage function).

### PR DESCRIPTION
**Pull request by steffnay**
This feature adds the ability to use a native Duplex stream for inserting rows into BigQuery via the `/insertAll` endpoint and reading the API response. Implements batching of rows via `RowBatch` and `RowQueue` classes.

Adds:

- Table.createInsertStream()
- RowQueue
- RowBatch

- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #506 🦕

### CODE REVIEW

**Review by feywind — COMMENTED** [review]
Sorry for the flood of comments ^^; I'm happy to talk over any of them and/or help fix some of the `any` comments if you like.

File: src/rowQueue.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+export const defaultOptions = {
+  // The maximum number of rows we'll batch up for insert().
+  maxOutstandingRows: 300,
+
+  // The maximum size of the total batched up rows for insert().
+  maxOutstandingBytes: 1 * 1024 * 1024,
+
+  // The maximum time we'll wait to send batched rows, in milliseconds.
+  maxDelayMillis: 10000,
```
[the remark this answers is not in the document]
Comment (steffnay) [reply src/rowQueue.ts:35]: Just based off similar batching of messages in nodejs-pubsub, do you have an idea of a timeframe that is better aligned for BigQuery inserts?

File: src/rowQueue.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type RowMetadata = any;
```
[the remark this answers is not in the document]
Comment (steffnay) [reply src/rowQueue.ts:39]: This is actually an existing type used for table inserts, I need to remove this replication and follow-up about changing it.

File: src/rowQueue.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  pending?: NodeJS.Timer;
+  constructor(table: Table, dup: Stream, options?: InsertStreamOptions) {
+    this.table = table;
+    this.stream = dup;
+    this.inFlight = false;
+
+    const opts = typeof options === 'object' ? options : {};
+
+    if (opts.insertRowsOptions) {
+      this.insertRowsOptions = opts.insertRowsOptions;
```
[the remark this answers is not in the document]
Comment (steffnay) [reply src/rowQueue.ts:83]: Yeah, that'd be great.

File: src/rowQueue.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+          (insertError: GoogleErrorBody) => {
+            return {
+              errors: insertError.errors!.map(error => {
+                return {
+                  message: error.message,
+                  reason: error.reason,
+                };
+              }),
+              // eslint-disable-next-line @typescript-eslint/no-explicit-any
+              row: rows[(insertError as any).index],
```
[the remark this answers is not in the document]
Comment (steffnay) [reply src/rowQueue.ts:183]: That's a good idea. Here I'm just copying the logic that exists in the `Table._insert` method for consistency, but I thin it'd be good to update these both separately in a type cleanup. https://github.com/googleapis/nodejs-bigquery/blob/f39833a9af3389c3877f470d8ed9766b5e3aabc3/src/table.ts#L2188

File: src/rowQueue.ts [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+   */
+  setOptions(options?: RowBatchOptions): void {
+    const defaults = {
+      maxBytes: defaultOptions.maxOutstandingBytes,
+      maxRows: defaultOptions.maxOutstandingRows,
+      maxMilliseconds: defaultOptions.maxDelayMillis,
+    };
+
+    // TODO: set individual defaults for any unset options
+    const opts = typeof options === 'object' ? options : defaults;
```
[the remark this answers is not in the document]
Comment (steffnay) [reply src/rowQueue.ts:219]: Ah, thank you! I couldn't remember what that syntax was!

File: test/rowBatch.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    it('should return true if it can fit', () => {
+      const canFit = batch.canFit(row);
+      assert.strictEqual(canFit, true);
+    });
+  });
+
+  describe('isAtMax', () => {
+    it('should return true if at max row limit', () => {
+      // tslint:disable-next-line ban
+      Array(1000)
```
[the remark this answers is not in the document]
Comment (steffnay) [reply test/rowBatch.ts:104]: I'm not sure either, but it seems to be okay without it now(?)

File: test/rowQueue.ts [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+    },
+    apiEndpoint: 'bigquery.googleapis.com',
+    request: util.noop,
+  },
+} as {}) as _root.Dataset;
+
+describe('Queues', () => {
+  const sandbox = sinon.createSandbox();
+
+  const Table = require('../src/table').Table;
```
[the remark this answers is not in the document]
Comment (steffnay) [reply test/rowQueue.ts:68]: that's definitely better.

File: src/rowQueue.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+ * @param {InsertStreamOptions} options Insert and batch options.
+ */
+export class RowQueue {
+  table: Table;
+  stream: Stream;
+  insertRowsOptions: InsertRowsOptions = {};
+  batch: RowBatch;
+  batchOptions?: RowBatchOptions;
+  inFlight: boolean;
+  pending?: NodeJS.Timer;
```
[the remark this answers is not in the document]
Comment (steffnay) [reply src/rowQueue.ts]: This is a great point, updated!

File: src/rowQueue.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (feywind) [reply src/rowQueue.ts:85]: Ah geez, I never actually gave this link to you... Here's one method:

https://github.com/googleapis/nodejs-pubsub/blob/38fba8bb98636d21c1d2005be4352fb8190e1813/src/publisher/index.ts#L279

File: src/rowQueue.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (feywind) [reply src/rowQueue.ts:181]: Sounds good :)

**Review by feywind — APPROVED** [review]
I can't remember what all I commented on now, but overall it looks good :)

File: src/rowQueue.ts [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+   */
+  setOptions(options?: RowBatchOptions): void {
+    const defaults = {
+      maxBytes: defaultOptions.maxOutstandingBytes,
+      maxRows: defaultOptions.maxOutstandingRows,
+      maxMilliseconds: defaultOptions.maxDelayMillis,
+    };
+
+    // TODO: set individual defaults for any unset options
+    // const opts = typeof options === 'object' ? options : defaults;
```
Comment (tswast) [inline src/rowQueue.ts:218]: Is this meant to be uncommented or removed?

**Review by tswast — APPROVED** [review]

File: src/rowQueue.ts (hunk 9 above)
Comment (steffnay) [reply src/rowQueue.ts:218]: Ah, meant to be updated per Megan's review. Done now!

**Pull request merged**